### PR TITLE
Issue#42 Write an ADR on our choices of database structure

### PR DESCRIPTION
**Pull request by JolanThomassin**
Name: **dev-rel-docs/adr/002-database-structure.md**

This ADR explains our plan to **organize our databases better**. You can find all the details in this [document ]( https://github.com/ai-cfia/louis-db/issues/24). In a nutshell, we're restructuring our databases to work together smoothly as we expand our product range. We'll have a **main repository named "ailab-db"** and add modules for specific products like "louis," "nachet," or "finesse." This approach offers flexibility and adaptability. The ADR aims to **clarify how** we're organizing everything **and why** we're doing it this way.

### CODE REVIEW

**Comment by JolanThomassin** [comment]
Now this is related to the main 👍

File: adr/002-database-structure.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+![AiLab-db](./002-database-structure-diagram.png)
+
+## Executive Summary
+This ADR discusses our choices for the ailab-db database structure. We want to encourage collaboration between various teams and products while increasing flexibility and ease of maintenance.
+
+## Context
+Our previous ailab-db architecture required maintenance of several databases, notably "louis-db" and "nachet-db." Although this approach initially provides clarity, it ultimately results in redundancy, less flexibility, and collaboration issues, which may cause data conflicts. We have a better way to maintain databases with this approach, which uses a Consolidated Database Approach and only uses one database.
+
+## Decision
+We have decided to transition to a Consolidated Database Approach for AiLab-db. This decision is rooted in our evaluation criteria, which emphasize flexibility, maintainability, efficiency, and collaborative potential.
```
Comment (rngadam) [inline adr/002-database-structure.md:13]: you supposed decided to write ailab-db AiLab-db. be consistent and use the lowercase name

File: adr/002-database-structure.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# ADR 002: ADR on our choices of database structure
+
+## Optional Illustration
+![AiLab-db](./002-database-structure-diagram.png)
+
+## Executive Summary
+This ADR discusses our choices for the ailab-db database structure. We want to encourage collaboration between various teams and products while increasing flexibility and ease of maintenance.
```
Comment (rngadam) [inline adr/002-database-structure.md:7]: We want to encourage data sharing between software products produced by the CFIA AI Lab while increasing flexibility and ease of maintenance by consolidating our product databases under a single database server.

**Review by rngadam — CHANGES_REQUESTED** [review]

File: adr/002-database-structure.md (hunk 1 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: The last Ailab-db has been found and fixed; now all Ailab-db entries are being written correctly. [8ce9ca3](https://github.com/ai-cfia/dev-rel-docs/pull/57/commits/8ce9ca30dca7be4581f889c58a878853e9c6bfd1)

File: adr/002-database-structure.md (hunk 2 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: I changed the text a little so it's more clear and it follow the text you give me in this comment. [a14acb8](https://github.com/ai-cfia/dev-rel-docs/pull/57/commits/a14acb86abec7bf8524cc9f97ba7f1215ae981c5)

File: adr/002-database-structure.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# ADR 002: ADR on our choices of database structure
+
+## Optional Illustration
```
Comment (melanie-fressard) [inline adr/002-database-structure.md:3]: Maybe change optional illustration to an actual title for your diagram

File: adr/002-database-structure.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+![AiLab-db](./002-database-structure-diagram.png)
+
+## Executive Summary
+This ADR describes our plan to promote sharing data between software products created by the CFIA AI Lab. At the same time, we aim to make things more flexible and easier to maintain by bringing all our product databases together in a single database server.
```
Comment (melanie-fressard) [inline adr/002-database-structure.md:7]: you need to say what CFIA and AI Lab stand for before using the acronyms

Comment (melanie-fressard) [reply adr/002-database-structure.md:7]: adr too

File: adr/002-database-structure.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+## Context
+Our previous ailab-db architecture required maintenance of several databases, notably "louis-db" and "nachet-db." Although this approach initially provides clarity, it ultimately results in redundancy, less flexibility, and collaboration issues, which may cause data conflicts. We have a better way to maintain databases with this approach, which uses a Consolidated Database Approach and only uses one database.
```
Comment (melanie-fressard) [inline adr/002-database-structure.md:10]: you need to wrap your text at column 80 with the extension Rewrap

File: adr/002-database-structure.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+## Decision
+We have decided to transition to a Consolidated Database Approach for ailab-db. This decision is rooted in our evaluation criteria, which emphasize flexibility, maintainability, efficiency, and collaborative potential.
```
Comment (melanie-fressard) [inline adr/002-database-structure.md:13]: Maybe explain a bit more what is Consolidated Database Approach

File: adr/002-database-structure.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+Cons:
+
+- **Significant effort required to update files:** If you don't initaly have a consolidated database structure, transitioning involves renaming and moving files, which can be a labor-intensive process. It may require a substantial initial effort to reorganize and migrate data.
+
+- **Development process complexity:** Adapting to the new database structure may introduce complexity to the development process. Teams may need to adjust their workflows and applications to accommodate the changes, potentially leading to temporary challenges. This required the team to have a good planning or prototype about schema and clear relationships between databases.
+
+## Consequences:
+
+### Positive Consequences:
```
Comment (melanie-fressard) [inline adr/002-database-structure.md:58]: don't separate both, you should present negative aspects here but nuance it with positive ones

File: adr/002-database-structure.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+- Collaboration will be better because teams will operate more smoothly and promote stronger cross-functional cooperation.
+- By eliminating redundant data, the single database will produce information that is more accurate and dependable.
+
+### Negative Consequences:
+
+- Renaming and moving files will necessitate a significant initial effort.
+- Adapting to the new structure may temporarily complicate the development process.
```
Comment (melanie-fressard) [inline adr/002-database-structure.md:66]: you don't have the reference section, I don't think you need one but it should be here anyway in my opinion

**Review by melanie-fressard — CHANGES_REQUESTED** [review]
great job, I think it needs some changes but it should be good once it's done imo

File: adr/002-database-structure.md (hunk 4 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: Yeah Ricky told me to do that and I forgot, it's done now

File: adr/002-database-structure.md (hunk 3 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: I changed it, but I'm not sure about the results

File: adr/002-database-structure.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
+# ADR 002: ADR on our choices of database structure
```
Comment (rngadam) [inline adr/002-database-structure.md:1]: # ADR-002: database structure design

File: adr/002-database-structure.md (hunk 5 above)
Comment (rngadam) [reply adr/002-database-structure.md:14]: I don't see the document wrapped?

File: adr/002-database-structure.md (hunk 8 above)
Comment (rngadam) [reply adr/002-database-structure.md:69]: please research a bit what good industry practices for postgresql / databases

**Review by rngadam — CHANGES_REQUESTED** [review]

**Review by melanie-fressard — APPROVED** [review]

File: adr/002-database-structure.md [hunk 10]
```diff
@@ -67,3 +67,16 @@ Cons:
 
 - Renaming and moving files will necessitate a significant initial effort.
 - Adapting to the new structure may temporarily complicate the development process.
+
+### References
+
+Any documentation, research, or data that support the decision can be cited here. This can include academic papers, blog posts, or internal documents.
+
+- <a id="ref-matillion"></a>What is Data Consolidation? (2020, December 16). Matillion.
+  https://www.matillion.com/blog/what-is-data-consolidation
```
Comment (rngadam) [inline adr/002-database-structure.md:77]: could we perhaps extract key quotes from these references into our doc and have that quote linked to the reference here?

File: adr/002-database-structure.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
+# ADR-002: database structure design
+
+## CFIA AI Lab Database Diagram
+![AiLab-db](./002-database-structure-diagram-organization.png)
+![AiLab-db](./002-database-structure-diagram.png)
```
Comment (rngadam) [inline adr/002-database-structure.md:5]: probably need to finish ADR-008 first so we can reference it here.

Could you try to use ChatGPT image description feature to provide a better alt-text to your images/diagrams?

**Review by rngadam — CHANGES_REQUESTED** [review]

File: adr/002-database-structure.md (hunk 10 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md:77]: That's a good idea, I'll work on that

File: adr/002-database-structure.md (hunk 11 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md:5]: I don't have this functionality , but I'll ask someone to to that for me.

**Comment by rngadam** [comment]
@JolanThomassin this is still pending changes, we should get it fixed and have the team review it.

**Review by JolanThomassin — COMMENTED** [review]
Here is the needed modification, I'm going to add the other for review now.

File: adr/002-database-structure.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+### Negative Consequences:
+
+- Renaming and moving files will necessitate a significant initial effort.
+- Adapting to the new structure may temporarily complicate the development process.
+
+### References
+
+Any documentation, research, or data that support the decision can be cited here. This can include academic papers, blog posts, or internal documents.
+
+- [Toptal - Database Design Bad Practices: Are You Making These Mistakes?](https://www.toptal.com/database/database-design-bad-practices)
```
Comment (k-allagbe) [inline adr/002-database-structure.md:77]: We agreed to use APA for references. Check https://github.com/ai-cfia/dev-rel-docs/blob/main/adr/001-ailab-using-adr.en-ca.md#format

**Review by k-allagbe — CHANGES_REQUESTED** [review]
Don't forget to update the `index.md` file. A rebase might be needed for that.

File: adr/002-database-structure.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+Cons:
+
+- **Redundancy:** Maintaining separate databases can lead to data redundancy. If the same data needs to be used across multiple products or teams, it may be duplicated in multiple databases, which can lead to inconsistencies and increased storage requirements.
+
+- **Reduced flexibility:** Having separate databases for each product or team can reduce flexibility. If there is a need to share or integrate data across products or teams, it may require complex data synchronization processes.
+
+- **Collaboration challenges:** Collaboration between different teams or products may be challenging with separate databases. It can be harder to share and access data seamlessly, leading to potential communication and coordination issues.
+
+- **Potential data conflicts:** Separate databases may result in data conflicts, especially when different teams or products need to update the same data simultaneously. These conflicts can lead to data inconsistencies and errors. (Source: [Toptal - Database Design Bad Practices](https://www.toptal.com/database/database-design-bad-practices))
```
Comment (RussellJimmies) [inline adr/002-database-structure.md:41]: In-text citations should be linked to the detailed citations in the References section.
 
image

**Review by RussellJimmies — CHANGES_REQUESTED** [review]

File: adr/002-database-structure.md (hunk 11 above)
Comment (rngadam) [reply adr/002-database-structure.md]: I tried with the builtin Windows Narrator but was defeated by its complicated keybindings and navigation.

File: adr/002-database-structure.md (hunk 9 above)
Comment (rngadam) [reply adr/002-database-structure.md]: Sorry, should be:

# ADR-002: Database Structure Design

File: adr/002-database-structure.md (hunk 12 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: I changed it for APA, like in the ADR-005.

File: adr/002-database-structure.md (hunk 13 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: Now it's a reference to the end of the page reference.

**Review by JolanThomassin — COMMENTED** [review]
Now it should follow the other one on reference.

File: adr/002-database-structure.md (hunk 12 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: What is `index.md`?

File: adr/002-database-structure.md (hunk 9 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: It make more sense, modification done! [a91a2dc](https://github.com/ai-cfia/dev-rel-docs/pull/57/commits/a91a2dccff328aaddc346d69e72bbd20d956783e)

**Review by rngadam — CHANGES_REQUESTED** [review]
document is not wrapping at 80

File: adr/002-database-structure.md (hunk 12 above)
Comment (k-allagbe) [reply adr/002-database-structure.md]: This: https://github.com/ai-cfia/dev-rel-docs/blob/main/adr/index.md

Comment (k-allagbe) [reply adr/002-database-structure.md]: This: https://github.com/ai-cfia/dev-rel-docs/blob/main/adr/index.md

Comment (k-allagbe) [reply adr/002-database-structure.md]: This: https://github.com/ai-cfia/dev-rel-docs/blob/main/adr/index.md

**Review by k-allagbe — CHANGES_REQUESTED** [review]

**Review by ThomasCardin — APPROVED** [review]

**Review by redat97 — APPROVED** [review]
nice work !

File: adr/002-database-structure.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
... three green rectangles connected to the server, representing individual databases: "finance (db)" on the left, "ailab (db)" in the middle, and "science (db)" on the right. Further down, under the "ailab (db)" database, there are two additional green rectangles, labeled "nachet_v1.0.0 (schema)" and "louis_v1.0.0 (schema)", suggesting these are specific schemas or structures within the AI Lab database. Below the "nachet_v1.0.0 (schema)" there is another green rectangle labeled "score (view)", indicating a database view, and under the "louis_v1.0.0 (schema)" are two rectangles: "html_content (table)" and "crawl (table)", suggesting these are tables within the Louis schema. The diagram features the same email addresses as the previous one: "cfia.ai-ia.acia@inspection.gc.ca" and "jolan.thomassin@inspection.gc.ca". The date "2023-11-09" is also present, indicating when the diagram was likely created or updated. In the bottom right corner, there is the logo of the Canadian Food Inspection Agency, its French name "Agence canadienne d'inspection des aliments", and the Canadian flag, which signifies the national scope and authority of the organization.](./002-database-structure-diagram.png)
```
Comment (rngadam) [inline adr/002-database-structure.md:5]: Note that the diagram also wrongly uses _v

**Review by rngadam — CHANGES_REQUESTED** [review]

**Review by RussellJimmies — APPROVED** [review]

File: adr/002-database-structure.md (hunk 12 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md]: I'll add it after the merge

File: adr/002-database-structure.md (hunk 14 above)
Comment (JolanThomassin) [reply adr/002-database-structure.md:5]: The diagram was made before the decision, I updated it, now it follow the good version system.

**Review by k-allagbe — APPROVED** [review]

**Comment by rngadam** [comment]
> document is not wrapping at 80

@JolanThomassin

**Comment by JolanThomassin** [comment]
> > document is not wrapping at 80
> 
> @JolanThomassin

Yeah I see it and done the change, but for some reason I'm unable to push on it for now. I'll try to fix that ASAP.

image

**Comment by JolanThomassin** [comment]
I still have a problem and am not sure why it's happening on my VS Code. However, it seems that my commit was pushed, so everything should be fine! [e21bbe0](https://github.com/ai-cfia/dev-rel-docs/pull/57/commits/e21bbe05294392515a4f99cd6c13dfc5ad5fbf32)

**Review by rngadam — CHANGES_REQUESTED** [review]

File: adr/002-database-structure.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+
+## Executive Summary
+This Architecture Decision Record (ADR) describes our plan to promote sharing
+data between software products created by the Canadian Food Inspection Agency
+(CFIA) Artificial Intelligence (AI) Lab. At the same time, we aim to make things
+more flexible and easier to maintain by bringing all our product databases
+together in a single database server.
+
+## Context
```
[the remark this answers is not in the document]
Comment (JolanThomassin) [reply adr/002-database-structure.md:57]: I have added these 3 ones, and looking if other term need to be explain

File: adr/002-database-structure.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+Our previous ailab-db architecture required maintenance of several databases,
```
[the remark this answers is not in the document]
Comment (JolanThomassin) [reply adr/002-database-structure.md:58]: If we keep using ailab-db for every project, keeping this name sounds good to me. I'm not sure which application you're referring to when you mention the "app schema."

[the remark this answers is not in the document]
Comment (rngadam) [reply adr/002-database-structure.md:58]: what is the use of what we call  louis right now? mostly crawled documents. 

so the schema would be more accurately named docs, and could be used by any number of apps that require to ETL documents  from any source.

File: adr/002-database-structure.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (rngadam) [reply adr/002-database-structure.md:57]: Refer back to our technology for the definitions as those you provided are too generic to be helpful: 

https://www.postgresql.org/docs/current/glossary.html

**Review by rngadam — CHANGES_REQUESTED** [review]

**Review by rngadam — CHANGES_REQUESTED** [review]
there's one conversation thread that is unresolved @JolanThomassin

**Comment by JolanThomassin** [comment]
Indeed, some comments were still pending.  

image  

Now everything should be answered or resolved.

**Comment by rngadam** [comment]
@JolanThomassin let's discuss in-person future refactoring of the database to organize by entity types instead of products:

* docs
* people
* logs
* blobs

although the starting point is one db per product, we may want to end up at databases at high-level entities

**Review by rngadam — APPROVED** [review]
As discussed, we will revisit as we rethink consolidation of databases in a new PR

**Pull request merged**